bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(link_descr->flags, BGP_NLRI_TLV_LINK_DESCR_MT_ID)) {
+		size = bgp_linkstate_nlri_mtid_display(&link_descr->mtid, buf,
+						       size, first);
+		buf += strlen(buf);
+	}
+
+	if (!multiline) {
+		snprintf(buf, size, "}");
+		size -= strlen(buf);
+	}
+
+	return size;
+}
+
+static uint16_t
+bgp_linkstate_nlri_link_prefixlen(struct bgp_linkstate_type_link *p_link)
+{
+	return sizeof(struct bgp_linkstate_type_link) +
+	       p_link->link_descr.mtid.length;
+}
+
+static int bgp_linkstate_nlri_link_decode(struct bgp_linkstate_type_link **pp,
+					  uint8_t *pnt, uint16_t len)
+{
+	int ret = BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+	struct bgp_linkstate_type_link *p_link = *pp;
+	uint8_t *lim = pnt + len;
+	uint16_t type, length, *dst, *src;
+
+	p_link->proto = pnt_decode8(&pnt);
+	p_link->identifier = pnt_decode64(&pnt);
+
+	while (pnt < lim) {
+		type = pnt_decode16(&pnt);
+		length = pnt_decode16(&pnt);
+
+		if (!bgp_ls_tlv_check_size(type, length))
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+
+		switch (type) {
+		case BGP_LS_TLV_LOCAL_NODE_DESCRIPTORS:
+			if (p_link->local_node_descr.flags != 0)
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_link->local_node_descr.flags,
+				 BGP_NLRI_TLV_NODE_DESCR_LOCAL_NODE);
+			ret = bgp_linkstate_nlri_node_descriptor_decode(
+				&p_link->local_node_descr, pnt, length);
+			break;
+		case BGP_LS_TLV_REMOTE_NODE_DESCRIPTORS:
+			if (p_link->remote_node_descr.flags != 0)
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_link->remote_node_descr.flags,
+				 BGP_NLRI_TLV_NODE_DESCR_REMOTE_NODE);
+			ret = bgp_linkstate_nlri_node_descriptor_decode(
+				&p_link->remote_node_descr, pnt, length);
+			break;
+		case BGP_LS_TLV_LINK_LOCAL_REMOTE_IDENTIFIERS:
+		case BGP_LS_TLV_IPV4_INTERFACE_ADDRESS:
+		case BGP_LS_TLV_IPV4_NEIGHBOR_ADDRESS:
+		case BGP_LS_TLV_IPV6_INTERFACE_ADDRESS:
+		case BGP_LS_TLV_IPV6_NEIGHBOR_ADDRESS:
+			bgp_linkstate_nlri_link_descriptor_decode(
+				&p_link->link_descr, pnt, type, length);
+			break;
+		case BGP_LS_TLV_MULTI_TOPOLOGY_ID:
+			if (CHECK_FLAG(p_link->link_descr.flags,
+				       BGP_NLRI_TLV_PREFIX_DESCR_MT_ID))
+				/* TLV must be present only once */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			if ((length % 2) != 0)
+				/* Length must be a multiple of 2 */
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			SET_FLAG(p_link->link_descr.flags,
+				 BGP_NLRI_TLV_LINK_DESCR_MT_ID);
+			*pp = XREALLOC(MTYPE_BGP_LS_TYPE_LINK, *pp,
+				       sizeof(struct bgp_linkstate_type_link) +
+					       length);
+			p_link = *pp;
+			p_link->link_descr.mtid.length = length;
+			dst = (uint16_t *)&p_link->link_descr.mtid.data;
+			src = (uint16_t *)pnt;
+			for (int i = 0; i < (length / 2); i++)
+				dst[i] = ntohs(src[i]);
+			break;
+		default:
```

Comment:
unknown TLVs should be kept as is. TLv includes the length, so let us move the pointer.